plugins/hls-semantic-tokens-plugin/src/Ide/Plugin/SemanticTokens/Query.hs

```diff
@@ -0,0 +1,126 @@
+            let typeTokenType = fmap (typeSemantic hieKind) $ listToMaybe $ mapMaybe (identType . snd) spanInfos
+            let contextInfoTokenType = contextInfosMaybeTokenType infos
+            maximum <$> NE.nonEmpty (catMaybes [typeTokenType, contextInfoTokenType])
+
+        contextInfosMaybeTokenType :: Set.Set ContextInfo -> Maybe SemanticTokenType
+        contextInfosMaybeTokenType details = case NE.nonEmpty $ Set.toList details of
+            Just infos -> Just $ maximum $ NE.map infoTokenType infos
+            Nothing    -> Nothing
+
+-----------------------------------
+-- * extract location from HieAST a
+-----------------------------------
+
+hieAstSpanNames :: UniqSet Name -> HieAST a -> [(Range, Name)]
+hieAstSpanNames nameSet ast = if null (nodeChildren ast) then
+    getIds ast else concatMap (hieAstSpanNames nameSet) (nodeChildren ast)
+    where
+        -- getIds :: HieAST a -> [(Span, Name)]
+        getIds ast' = [(realSrcSpanToRange $ nodeSpan ast', c)
+                    | (Right c, d) <- Map.toList $ getNodeIds' ast
+                    , elemNameSet c nameSet
+                    -- at least get one info
+                    , not $ any isEvidenceBind $ identInfo d
+                    , not $ any isEvidenceUse $ identInfo d
+                    , Set.size (identInfo d) > 0
+                    -- some derived occName is visible (some fields)
+                    -- , not $ isDerivedOccName (occName c)
+                    ]
+        getNodeIds' :: HieAST a -> Map.Map Identifier (IdentifierDetails a)
+        getNodeIds' = Map.foldl' combineNodeIds Map.empty
+            .  Map.filterWithKey (\k _ -> k == SourceInfo)
+            . getSourcedNodeInfo . sourcedNodeInfo
+
+        combineNodeIds :: Map.Map Identifier (IdentifierDetails a)
+                                -> NodeInfo a -> Map.Map Identifier (IdentifierDetails a)
+        ad `combineNodeIds` (NodeInfo _ _ bd) = Map.unionWith (<>) ad bd
+
+
+-------------------------------------------------
+-- * extract semantic tokens from NameSemanticMap
+-------------------------------------------------
+
+semanticTokenAbsoluteSemanticTokens :: [SemanticTokenAbsolute] -> Either Text SemanticTokens
+semanticTokenAbsoluteSemanticTokens = makeSemanticTokens defaultSemanticTokensLegend . List.sort
+
+extractSemanticTokensFromNames :: NameSemanticMap -> [(Range, Name)] -> [SemanticTokenAbsolute]
+extractSemanticTokensFromNames nsm =
+    mapMaybe (uncurry toAbsSemanticToken) . mergeNameFromSamSpan . mapMaybe (getSemantic nsm)
+    where
+        -- merge all tokens with same span
+        -- mergeNameFromSamSpan :: [(Span, SemanticTokenType)] -> [(Span, SemanticTokenType)]
+        mergeNameFromSamSpan xs = Map.toList $ Map.fromListWith (<>) xs
+
+        toAbsSemanticToken :: Range -> SemanticTokenType -> Maybe SemanticTokenAbsolute
+        toAbsSemanticToken (Range (Position startLine startColumn) (Position _endLine endColumn)) tokenType =
```

Comment:
Something like this might be a useful helper to add to `lsp` :thinking:

Reply:
Indeed, a more friendly maker function for lsp's semantic token is surely helpful.